pkg/agent/client.go

```diff
@@ -76,33 +85,154 @@ type AgentClient struct {
+	serviceAccountTokenPath string
 }
 
 func newAgentClient(address, agentID string, cs *ClientSet, opts ...grpc.DialOption) (*AgentClient, error) {
-	stream, err := NewRedialableAgentClient(address, agentID, cs, opts...)
+	a := &AgentClient{
+		cs:                      cs,
+		address:                 address,
+		agentID:                 agentID,
+		opts:                    opts,
+		probeInterval:           cs.probeInterval,
+		stopCh:                  make(chan struct{}),
+		serviceAccountTokenPath: cs.serviceAccountTokenPath,
+		connManager:             newConnectionManager(),
+	}
+	err := a.Connect()
 	if err != nil {
 		return nil, err
 	}
-	return newAgentClientWithRedialableAgentClient(stream), nil
+	return a, nil
 }
 
-func newAgentClientWithRedialableAgentClient(rac *RedialableAgentClient) *AgentClient {
-	return &AgentClient{
-		connManager: newConnectionManager(),
-		stream:      rac,
-		stopCh:      rac.stopCh,
+// Connect makes the grpc dial to the proxy server. It returns the serverID
+// it connects to.
+func (a *AgentClient) Connect() error {
+	conn, err := grpc.Dial(a.address, a.opts...)
+	if err != nil {
+		return err
+	}
+	ctx := metadata.AppendToOutgoingContext(context.Background(), header.AgentID, a.agentID)
+	if a.serviceAccountTokenPath != "" {
+		if ctx, err = a.initializeAuthContext(ctx); err != nil {
+			conn.Close()
+			return err
+		}
+	}
+	stream, err := agent.NewAgentServiceClient(conn).Connect(ctx)
+	if err != nil {
+		conn.Close()
+		return err
+	}
+	serverID, err := serverID(stream)
+	if err != nil {
+		conn.Close()
+		return err
+	}
+	serverCount, err := serverCount(stream)
+	if err != nil {
+		conn.Close()
+		return err
 	}
+	a.conn = conn
+	a.stream = stream
+	a.serverID = serverID
+	klog.Infof("Connect to server %s", serverID)
+	a.serverCount = serverCount
+	return nil
 }
 
-// Close closes the underlying stream.
+// Close closes the underlying connection.
 func (a *AgentClient) Close() {
-	if a.stream == nil {
-		klog.Warning("Unexpected empty AgentClient.stream")
-		return
+	if a.conn == nil {
+		klog.Warning("Unexpected empty AgentClient.conn")
 	}
-	a.stream.Close()
+	a.conn.Close()
+	close(a.stopCh)
+}
+
+func (a *AgentClient) Send(pkt *client.Packet) error {
+	a.sendLock.Lock()
+	defer a.sendLock.Unlock()
+
+	err := a.stream.Send(pkt)
+	// TODO: add metrics for send err
+	if err != nil && err != io.EOF {
+		a.cs.RemoveClient(a.serverID)
+	}
+	return err
+}
+
+func (a *AgentClient) Recv() (*client.Packet, error) {
+	a.recvLock.Lock()
+	defer a.recvLock.Unlock()
+
+	pkt, err := a.stream.Recv()
+	// TODO: add metrics for recv err
+	if err != nil && err != io.EOF {
+		a.cs.RemoveClient(a.serverID)
+	}
+	return pkt, err
+}
+
+func serverCount(stream agent.AgentService_ConnectClient) (int, error) {
+	md, err := stream.Header()
+	if err != nil {
+		return 0, err
+	}
+	scounts := md.Get(header.ServerCount)
+	if len(scounts) == 0 {
```

Comment:
When is it ok to get more than 1 ServerCount? If you get more than 1, how do you know which one is the right value?

Reply:
Updated it to check the length is 1.